InceptionV3 MIGraphX + Onnxrt End to End Example

TedThemistokleous opened the pull request:
skip ci
Reusing the changes from resnet50 to get this running and get a result. This is based off the resnet50 changeset

This is a piece that allows us to reuse code as an example of Onnxruntime inference for inceptionV3 and then again in DLM testing.

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+    parser.add_argument(
+        "--image_dir",
+        required=False,
+        default="../resnet50/images",
```
Is this default value value valid when running this example standalone ?

TedThemistokleous replied:
I figured I'd share images between the two so a user can compare

kahmed10 replied:
we need to change the directory structure in this case. I suggest making a `img_samples` directory that can be shared between the examples.

TedThemistokleous replied:
Will do

TedThemistokleous replied:
Done

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+        print("Reading in Imagenet classes")
+
+    # Read the categories
+    with open("./imagenet_classes.txt", "r") as f:
```
Is this assumed to be available ?

TedThemistokleous replied:
yes. Will restructure folder here to put images and image_classes.txt in an images directory shared between this and resnet50 examples

TedThemistokleous replied:
Should we just pull this down/store here as well? Its shared between both examples

shivadbhavsar replied:
I think this should exist with the dataset, all the commonly available models use the same class mappings when training on imagenet as far as I know

TedThemistokleous replied:
Done

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+
+def softmax(x):
+    """Compute softmax values for each sets of scores in x."""
+    e_x = np.exp(x - np.max(x))
```
subtraction with `np.max(x)` is it approximation to avoid overflows/underflows ?

TedThemistokleous replied:
Thats what I thought too. I took this from an example when I did resnet50. If its an issue I can change

umangyadav reviewed (changes requested):
add README.md file on what this example is about, how to run and what options are available to users etc.

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+
+    session_ops = onnxruntime.SessionOptions()
+    if flags.verbose:
+        session_ops.log_verbosity_level = 0
+        session_ops.log_severity_level = 0
```
level = 0 does it print anything ? isn't it default ?

TedThemistokleous replied:
Default is 2. Onnxruntime specifies verbosity levels and severity here: https://onnxruntime.ai/docs/api/python/api_summary.html#onnxruntime.SessionOptions.log_severity_level

its meant to print and see where the nodes go and how the model is invoked along with any other events through Onnxruntime. We can still enable MIGRAPHX_TRACE_EVAL to verify that the script does infact run as fp16 or fp32 eg

MIGRAPHX_TRACE_EVAL=1 python3 inceptionv3.py

umangyadav replied:
I meant for `verbosity_level = 0` not severity

TedThemistokleous replied:
Its so we ensure the verbosity_level=0. I don't want to assume otherwise/defaults.

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+
+        input_image = Image.open(str(flags.image_dir + "/" + img))
+        preprocess = T.Compose([
+            T.Resize(342),
```
Model i think takes 299x299 already. Is resize necessary ?

TedThemistokleous replied:
Yes. its based on the requirements specified by the pytorch example docs:

https://pytorch.org/vision/main/models/generated/torchvision.models.inception_v3.html#torchvision.models.inception_v3

TedThemistokleous replied:
Fair I can remove it.

TedThemistokleous replied:
Done, remove crop + resize

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+
+    if flags.verbose:
+        print("Running Complete")
+        print(latency)
```
```suggestion
       print("Total time :", latency)
```

TedThemistokleous replied:
Done

umangyadav replied:
I don't see this change for some reason

TedThemistokleous replied:
Batched stuff before I pushed. should be gone now

umangyadav replied:
I still don't see it

TedThemistokleous commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+
+    batch_size = 0
+    for img in fileList:
+        filename = img  # change to your filename
```
will clean up

TedThemistokleous replied:
Done

umangyadav replied:
did you miss pushing those changes ? I don't see it cleaned up

TedThemistokleous replied:
I pushed now

shivadbhavsar commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+    e_x = np.exp(x - np.max(x))
+    return e_x / e_x.sum()
+
+def run_sample(session, categories, latency, inputs, topk, batch_size):
```
This method will include the host-device copy overheads. There is some [documentation](https://onnxruntime.ai/docs/api/python/api_summary.html#data-on-device:~:text=The%20input%20data%20and%20output%20data%20are%20both%20on%20a%20device%2C%20users%20directly%20use%20the%20input%20and%20also%20place%20output%20on%20the%20device) for leaving intputs and outputs on host.

TedThemistokleous replied:
Ah I thought timing would not include that and the copy happens after the bind.

shivadbhavsar replied:
ahh i missed that explicit output copy. Rereading the docs, Im not sure this will include the host->device copy for the input though. Also, will python actually wait for `session.run_with_iobinding(io_binding)` to get a result from device before moving to the next line?
Im concerned that it might be executed async and python will only wait on the line that actually uses this output.

shivadbhavsar replied:
I'd say just verify youre getting reasonable values for this (compared to driver perf) and feel free to resolve this

TedThemistokleous replied:
I don't think its getting run async as I'm 

inception_v3, Rate = 107.53 QPS

vs our driver

Batch size: 1
Rate: 218.791 inferences/sec
Total time: 4.57057ms

Oddly we're slower. Maybe I am grabbing the time to transfer here as well

shivadbhavsar replied:
ok its unlikely that you'll see it drop to half just because of transfers. Look at my other comment about averaging. And also this is probably unreliable because you're measuring latency on a "cold start run". Even when you average, make sure to run it once (or maybe even n times, Im not sure about the inner workings of ort) before recording the measurements

TedThemistokleous replied:
Yeah I think its cold start.

inception_v3, Rate = 206.74 QPS

After doing an additional run before taking the measurement. Maybe its worth tracking setup time between runs.?

TedThemistokleous replied:
I added a loop to run this 100 times like how we do a migraphx-driver perf. Looks like it gets a better measurement slightly slower (due to the overhead) than our migraphx-driver perf runs on the model.

shivadbhavsar commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+    io_binding = session.io_binding()
+    io_binding.bind_cpu_input('input', inputs.cpu().detach().numpy())
+    io_binding.bind_output('output')
+    start = time.time()
```
In general when you want measure execution time, it should be done as an average over many iterations.  I think this function is doing too much, maybe just let it compute the predictions and print them, but take the benchmarking to another function that does this and also eliminates the data copy overheads?

TedThemistokleous replied:
I think this is handled correctly now with a loop. I'm seeing a small drop off between Onnxrt and our driver about 3-10 QPS between the two

TedThemistokleous commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+import subprocess
+
+#Use most upto date weights
+inception_v3 = models.inception_v3(weights=models.Inception_V3_Weights.DEFAULT, progress=True)
```
I'll add the eval() back in. I know I removed this earlier.

TedThemistokleous replied:
Done

TedThemistokleous commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+    image_width = 299
+    x = torch.randn(flags.batch, 3, image_height, image_width, requires_grad=True)
+    #torch_out = inception_v3(x)
+    torch.onnx.export(
```
Oh nice. yeah I see it. I can add it. 

For ref: https://pytorch.org/docs/stable/onnx_torchscript.html#torch.onnx.export

TedThemistokleous replied:
Added

shivadbhavsar commented:
> add README.md file on what this example is about, how to run and what options are available to users etc.

This, and also a `requirements.txt` for all the python deps

TedThemistokleous commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,236 @@
+            print("Preprocess: " + img)
+
+        input_image = Image.open(str(flags.image_dir + "/" + img))
+        preprocess = T.Compose([
```
Ah didn't realize you could do that. I guess I should do the same for the resnet50 setup too then?

shivadbhavsar replied:
Ya its probably to just use the input transforms they provide in the API

TedThemistokleous replied:
done

TedThemistokleous commented:
> > add README.md file on what this example is about, how to run and what options are available to users etc.
> 
> This, and also a `requirements.txt` for all the python deps

Ive added an preq_steps.sh to run but I can add a requirements.txt too

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
````diff
@@ -0,0 +1,78 @@
+```
+$ cd AMDMIGraphX
+$ docker build -t migraphx .
+$ docker --device='/dev/kfd' --device='/dev/dri' <your docker settings and mounts> --group-add video -it migraphx
````
```suggestion
$ docker run --device='/dev/kfd' --device='/dev/dri' <your docker settings and mounts> --group-add video -it migraphx
```

TedThemistokleous replied:
Done

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+- [Example output](#example-Output)
+
+## Basic Setup
+Before running inference we must first install MIGraphX via the Docker method as it also downloads onnxruntime into the dockerfile created. 
```
Doesn't necessaily need to build docker from MIGraphX. Users can use any docker with rocm and install migraphx inside it.

TedThemistokleous replied:
Said just "install MIGraphX" here instead.

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
````diff
@@ -0,0 +1,78 @@
+
+```
+$ 
+$ cd examples/onnxruntime/invepctionv3
````
spelling 
```suggestion
$ cd examples/onnxruntime/inceptionv3
```

TedThemistokleous replied:
done

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
````diff
@@ -0,0 +1,78 @@
+Once the docker file has been installed and you're inside the folder run
+
+```
+$ rbuild develop -d deps -B build
+$ cd build 
+$ make -j$(nproc) package && dpkg -i *.deb
````
Is there a way to not build from source and run this example ? It is too much for an example.

TedThemistokleous replied:
Until we get Onnxruntime wheels there isn't a way around building here. I don't have a URL for the target wheel files yet.

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+
+For the tools (scope.jpg and screwdrivers.jpg) these are both in-class images
+
+For the bird.jpg image, the imagenet_classes.txt generated doesn't contain that of a cockatiel and thus the model attempts to find the closet 
```
```suggestion
For the bird.jpg image, the imagenet_classes.txt generated doesn't contain that of a cockatiel and thus the model attempts to find the closest 
```

TedThemistokleous replied:
Done

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+For the tools (scope.jpg and screwdrivers.jpg) these are both in-class images
+
+For the bird.jpg image, the imagenet_classes.txt generated doesn't contain that of a cockatiel and thus the model attempts to find the closet 
+match to the animal found in the image
+
+using bird.jpg. Image of an oscilliscope which is in the imagenet class labels
```
Formatting looks incorrect here

TedThemistokleous replied:
Done

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Description
+
+This example demonstrates how to perform an MIGraphX Python API invepctionv3 inference through onnxruntime. The model used here is from Torchvision's pretrained invepctionv3 model
```
```suggestion
This example demonstrates how to perform an MIGraphX Python API invepctionv3 inference through onnxruntime. The model used here is from Torchvision's pretrained inceptionv3 model
```

TedThemistokleous replied:
Done

TedThemistokleous commented:
Added some extra filtering on processing step and cleaned up a few more things as I was retesting this with a larger dataset (found on our NAS imagenet/ILSVRC2012) for testing as some images were only using 1 channel instead of 3.

 This was the ask from QA Performance team to be able to do a comparison but should be a better way to handle input if a user decides to use a dataset

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Description
+
+This example demonstrates how to perform an MIGraphX Python API inceptionv3 inference through onnxruntime. The model used here is from Torchvision's pretrained invepctionv3 model
```
```suggestion
This example demonstrates how to perform an MIGraphX Python API inceptionv3 inference through onnxruntime. The model used here is from Torchvision's pretrained inceptionv3 model
```

umangyadav commented on examples/onnxruntime/inceptionV3/README.md:
```diff
@@ -0,0 +1,78 @@
+
+This directory contains everything needed to perform an inference once MIGraphX has been installed in the docker container
+Once you've build and installed MIGraphX as a deb package, go to the examples folder, run the pre-req script to build and install
+onnxruntime and then install the approrpaite version of pytorch from project root.
```
```suggestion
onnxruntime and then install the appropriate version of pytorch from project root.
```

umangyadav commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,276 @@
+
+        if verbose:
+            for catid in top5_catid:
+                print(categories[catid], output[catid])
```
Print some text here that indicates what it is printing otherwise it will get buried inside verbose logs.

umangyadav commented on examples/onnxruntime/inceptionV3/prereq_steps.sh:
```diff
@@ -0,0 +1,34 @@
+#####################################################################################
+
+#Install most recent stable version of pytorch
+pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/rocm5.6
```
Is there 5.7 or 6.0 pytorch available ?

TedThemistokleous replied:
Not publicly yet. During testing I have used the 6.0 by grabbing artifactory build links

shivadbhavsar replied:
yea you have to pull it from https://repo.radeon.com/rocm/manylinux/rocm-rel-6.0/

shivadbhavsar commented on examples/onnxruntime/inceptionV3/inceptionv3.py:
```diff
@@ -0,0 +1,276 @@
+        print("inception_v3, Rate = {} QPS".format(
+            format((((flags.batch)) / (sum(latency) / len(latency))), '.2f')))
+    else:
+        print("inception_v3, Total time = {} ms".format(
```
Call this "Average execution time" rather than "Total time", and in the calculation for both qps and time remove the first (cold start) measurement. You can simply replace `latency` with `latency[1:]`

TedThemistokleous replied:
Total time is to Mirror what we do with our driver but sure I can change it. I think there was a comment about setting it to Total time instead of just "latency". Average is better here though

TedThemistokleous replied:
Done

shivadbhavsar commented on examples/onnxruntime/resnet50/resnet50.py:
```diff
@@ -227,6 +264,7 @@ def main():
     if flags.QPS:
         print("resnet50, Rate = {} QPS ".format(
             format((((flags.batch) / (sum(latency) / len(latency)))), '.2f')))
```
Same as comment for inceptionv3, remove the cold start run from the average

TedThemistokleous replied:
Done

umangyadav reviewed (approved):
LGTM now
